src/viam/examples/modules/complex/main.cpp

```diff
@@ -0,0 +1,107 @@
+#include <memory>
+#include <signal.h>
+
+#include <boost/log/trivial.hpp>
+#include <grpcpp/grpcpp.h>
+#include <grpcpp/server_context.h>
+
+#include <viam/api/common/v1/common.grpc.pb.h>
+#include <viam/api/component/generic/v1/generic.grpc.pb.h>
+#include <viam/api/robot/v1/robot.pb.h>
+
+#include <viam/sdk/components/base/base.hpp>
+#include <viam/sdk/components/component.hpp>
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/module/module.hpp>
+#include <viam/sdk/module/service.hpp>
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/resource/resource.hpp>
+#include <viam/sdk/rpc/dial.hpp>
+#include <viam/sdk/rpc/server.hpp>
+
+#include "base/impl.hpp"
+#include "gizmo/api.hpp"
+#include "gizmo/impl.hpp"
+#include "summation/api.hpp"
+#include "summation/impl.hpp"
+
+using namespace viam::sdk;
+
+int main(int argc, char** argv) {
+    if (argc < 2) {
+        std::cerr << "Need socket path as command line argument" << std::endl;
+        return EXIT_FAILURE;
+    }
+    std::string socket_addr = argv[1];
+
+    // Use set_logger_severity_from_args to set the boost trivial logger's
+    // severity depending on commandline arguments.
+    set_logger_severity_from_args(argc, argv);
+    BOOST_LOG_TRIVIAL(debug) << "Starting module with debug level logging";
+
+    // C++ modules must handle SIGINT and SIGTERM. You can use the SignalManager
+    // class and its wait method to handle the correct signals.
+    SignalManager signals;
+
+    API base_api = Base::static_api();
+    Model mybase_model("viam", "base", "mybase");
+
+    std::shared_ptr<ModelRegistration> mybase_mr = std::make_shared<ModelRegistration>(
+        ResourceType("Base"),
+        base_api,
+        mybase_model,
+        [](Dependencies deps, ResourceConfig cfg) { return std::make_unique<MyBase>(deps, cfg); },
+        MyBase::validate);
+
+    Registry::register_model(mybase_mr);
+
+    API gizmo_api = Gizmo::static_api();
+    Model mygizmo_model("viam", "gizmo", "mygizmo");
+    // Make sure to explicity register resources with custom APIs. Note that
+    // this must be done in `main` and not in resource implementation files due
+    // to order of static initialization.
+    Registry::register_resource(gizmo_api, Gizmo::resource_registration());
```

Comment:
Many thanks to @acmorrow for helping me out here. We cannot statically register these custom resources within their own implementation files as their gRPC service descriptors will not be added to the generated pool at that point in time. We _can_ do that within the regular C++ SDK because it is split between two libraries (`libviamapi` and `libviamsdk`) that load sequentially and allow registration of gRPC service descriptors to happen before registration of resources. Since there's only an executable here, we cannot guarantee the same ordering. I think for now, we'll have to suggest (through the comment I wrote here) that users explicitly register their new APIs from within the main function. I've filed [RSDK-5206](https://viam.atlassian.net/browse/RSDK-5206) to consider whether we even want to continue the existing pattern in the C++ SDK.